README.md
# downscale-satelliteLST
A python class for enhancing the spatial resolution of Land Surface Temperature (LST) raster data using statistical downscaling.
![](LST_vs_DLST_compr.gif)

## Description
This class implements the typical workflow of a statistical downscaling scheme for enhancing the spatial resolution of satellite-derived Land Surface Temperatures (LST). It uses [GDAL](https://gdal.org/python/) to perform the resampling of the raster data and [scikit-learn](https://scikit-learn.org/stable/) to build the regression models, by [stacking](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.StackingRegressor.html) a [Random Forest](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html?highlight=random%20forest#sklearn.ensemble.RandomForestRegressor), a [Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html?highlight=ridge#sklearn.linear_model.Ridge) and a [SVM](https://scikit-learn.org/stable/modules/classes.html?highlight=svm#module-sklearn.svm) into an ensemble regressor.

### Input Data:
1. **LST**: A single raster dataset with one or more bands (each band is a LST image).
2. **Predictors**: A single raster dataset with one or more bands (each band is a predictor).

The class does **not** require the two raster datasets to have the same SRS and Bounding  Box. The only requirement is the predictors to be **within** the bounds of the LST data.

### Class checks before downscaling the LST data:
If a LST band misses more than 40% of its pixels, then this band is discarded and no model is built. In addition, if a model achieves a R2 that is lower than 0.5, it is also discarded. These two thresholds can be changed using the setters `SetMissingPxlsThreshold()` and `SetR2Threshold()`, respectively.

### Output:
A dictionary with the Downscaled LST (DLST) data of all the non-discarded models. The **spatial resolution** and the **SRS** of the output data is that of the predictors.

To save the DLST data as a raster dataset use the class method `SaveDLSTasGeotiff()` and provide a savename (the savepath is the workdir).

## Usage
```python
from osgeo import gdal
from DownscaleSatelliteLST import DownscaledLST     # Import the class

# Make an instance of the class
data = DownscaledLST(
        LST=gdal.Open("inputLST.tif"),				 
        predictors=gdal.Open("LSTpredictors.tif"),   
        LST_noDataVal=-1000,						
        predictors_noDataVal=-1000,		
        workdir="./DLST_save_folder",
        )

# Change the number of parallel jobs to 4 (the default value is 1; to use all cores use -1)
data.SetNumberOfJobs(4)

# Change the R^2 threshold for discarding a model (the default value is 0.5)
data.SetR2Threshold(0.6)

# Downscale the LST data and apply the residual correction.
DLST = data.ApplyDownscaling(residual_corr=True)

# Get a list with the LST bands that have been downscaled.
# LST bands that miss more than 40% of their pixels 
# and regression models that achieve a R^2 below the
# R2-threshold are discarded.
bands = data.GetDLSTBandIndices(indexing_from_1=False)

# Save a report with the scores of all the non-discarded models
# The report is saved in workdir
data.GenerateReport()

# Export the DLST data as a compressed Geotiff file
# The geotiff file is saved in workdir 
data.SaveDLSTasGeotiff(savename="DLST.tif")
```
For a working example check the folder [example](example).

## Things to keep in mind:
- The recommended datatype for the input rasters is **float**.
- If the LST or the predictors contain any **water bodies** or **clouds**, then these pixels should be **NoData**.
- All the NoData pixels should have the **same** NoData value.
- The class builds a **"global" regression model** for each LST band. Hence, it should be used with data that cover an area of **limited extent**, e.g. a city with its surroundings.
- If the predictors are gapless, the algorithm will generate DLST data and for the cloud-covered areas. Handle them with caution.

## To Do
- Add a class for preparing the predictors.
- Add unittests

## License
This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.

## Reference
If you use this class please cite:

    @phdthesis{Sismanidis2018,
        author = {Sismanidis, Panagiotis},
        pages  = {154},
        school = {National Technical University of Athens},
        title  = {{Applying Computational Methods for Processing Thermal Satellite Images of Urban Areas}},
        type   = {PhD Dissertation},
        year   = {2018}
    }


CHANGELOG.md
# CHANGELOG

## v.1.1.1 - 5 July 2022

- `DownscaledLST()` should now be used with scikit-learn v.1.1.1.
- `requirements.txt` was updated and `environment.yml` was added.
- Removed unused imports. Fixed some typos.

## v.1.1.0 - 5 March 2020

- The ensemble model is now derived using a stacked regressor instead of a voting regressor. The estimator used for the stacking is an elasticNET regressor with built-in cross-validation.
- The ensemble model is now built using a random forest, a ridge regressor and a SVM. The adaboost and elasticNET regressors used in v.1.0.0 have been dropped.
- The class now uses random search with cross validation to fine-tune the hyperparameters of the 3 regressors and thus does not require the user to estimate them beforehand. The default number of hyperparameter searches is 60 but it can be changed using the setter `set_num_searches`.
- The class now transforms the input predictors to follow a normal distribution using sklearn's QuantileTransformer. The number of bins is uses are equal to 50% of the length of y.
- The setters `SetAdaBoostParams`, `SetRandomForestParams`, `SetElasticNetParams` and `SetRidgeRegrParams` have been deleted.
- The methods `set_num_searches` for setting how many hyperparameter sets that will be tested and `set_num_jobs` for setting the maximum number of parallel jobs (default is 1) have been addded.


test/context.py
# -*- coding: utf-8 -*-

import sys
import os

sys.path.insert(
    0, os.path.abspath(os.path.join(os.path.dirname(__file__), os.path.pardir))
)

import DownscaleSatelliteLST


use_function.py

if __name__ == "__main__":

    import platform
    from osgeo import gdal
    from DownscaleSatelliteLST import DownscaledLST

    if platform.system() == "Linux" or platform.system() == "Darwin":  # Darwin is Mac
        #predictors_fname = "/mnt/e/Work/MyCodeRepository/sandbox/lst-downscaling_forAGU/example/eco_20180821T132843pca_pred_stack.tif" #eco_20180821T132843pca_pred_stack_withAspGHSL.tif" 
        #lst_fname = "/mnt/e/Work/MyCodeRepository/sandbox/lst-downscaling_forAGU/example/lsalst_500scenes.tif" #lsalst_2018_1453scenes.tif"
        #savedir = "/mnt/e/Work/Projects/NASA UrbClim/GOES16 082018 LA data/Downscaling Results/AthensReRun_4ParamDistrs/500_scenes_rf+sgd"
        predictors_fname = "/mnt/e/Work/Projects/NASA UrbClim/GOES16 082018 LA data/data/predictors entire scene/funion_Smooth_v4.2.tif" 
        lst_fname = "/mnt/e/Work/Projects/NASA UrbClim/GOES16 082018 LA data/data/GOES LST/GOESLST_LA_201808082130.tif"
        savedir = "/mnt/e/Work/Projects/NASA UrbClim/GOES16 082018 LA data/Downscaling Results/Test_5/funion_Smooth_v4.2_corr"

    elif platform.system() == "Windows":
        predictors_fname = r"E:\Work\Projects\NASA UrbClim\GOES16 082018 LA data\Downscaling Results\Test_1\LA_predictors.tif"
        lst_fname= r"E:\Work\Projects\NASA UrbClim\GOES16 082018 LA data\Downscaling Results\Test 7/GOESLST_LA_201808082130_v3.tif"
        savedir = (
            r"C:\Users\User\Desktop\NEW_CODE"
        )
    else:
        print("Unable to determine which OS you are using")

    predictors=gdal.Open(predictors_fname).ReadAsArray()

    p = DownscaledLST(
        LST=gdal.Open(lst_fname),
        predictors=gdal.Open(predictors_fname),
        LST_noDataVal=-1000,
        predictors_noDataVal=-1000,
        workdir=savedir,
    )
    
    p.SetR2Threshold(0.10)
    p.SetNumberOfJobs(4)
    p.SetRandomSearchNumber(60)
    
    
    data = p.ApplyDownscaling(residual_corr=True)
    p.GenerateReport()
    print(p.GetDLSTBandIndices(indexing_from_1=True))
    p.SaveDLSTasGeotiff(savename="LAT_Sep09_16_DLST.tif")


example/example_using_SEVIRI_data.py
from osgeo import gdal
from DownscaleSatelliteLST import DownscaledLST

# Raster filenames
lst_fname = "LSALST_20180819_Athens_15min.tif"
predictors_fname = "LST_predictors_100m.tif"


def main():

    # Read the raster data as gdal datasets
    Seviri_LST = gdal.Open(lst_fname)
    predictors_100m = gdal.Open(predictors_fname)

    # Make an instance of the downscaling class
    data = DownscaledLST(
            LST=Seviri_LST,
            predictors=predictors_100m,
            LST_noDataVal=-80,
            predictors_noDataVal=-1000,
            workdir="./Results",
        )

    # Change the number of parallel jobs to 4
    data.SetNumberOfJobs(4)

    # Set the R2 threshold for discarding
    # poorly-performing regression models.
    data.SetR2Threshold(0.5)

    # Apply the Downscaling
    data.ApplyDownscaling(residual_corr=True)

    # Save the dowscaled data as a geotiff file
    data.SaveDLSTasGeotiff(savename="downscaled_LST.tif")
    
    # Save the regression metrics in a txt file
    data.GenerateReport()

    # Get the indices of the non-discarded downscaled data
    print("LST bands that have been downscaled:")
    print(data.GetDLSTBandIndices(indexing_from_1=False))

if __name__ == "__main__":
    main()



example/DLST_SEVIRI_example.py
import platform
from osgeo import gdal
from DownscaleSatelliteLST import DownscaledLST

if platform.system() == "Linux" or platform.system() == "Darwin":  # Darwin is Mac
    #lst_fname = "/Users/panosis/Dropbox/MyCodeRepository/sandbox/Larissa_LST_UTM34N_120m_clp_v3.tif"
    #predictors_fname = "/Users/panosis/Dropbox/MyCodeRepository/sandbox/Larissa_B5_UTM34N_30m_clp_v2.tif"
    lst_fname = "/Users/panosis/Dropbox/MyCodeRepository/lst-downscaling/example/lsalst_2018_0809.tif"
    predictors_fname = "/Users/panosis/Dropbox/MyCodeRepository/lst-downscaling/example/eco_20180821T132843pca_pred_stack.tif"
    #lst_fname = "/Users/panosis/Desktop/DLST_tests/lC8_thess_500m.tif"
    #predictors_fname = "/Users/panosis/Desktop/DLST_tests/LC08_184032_20190508_LST.LST.tif"
    savedir = "/Users/panosis/Desktop/larissa_test5"

elif platform.system() == "Windows":
    lst_fname = r"C:\Users\User\Dropbox\MyCodeRepository\lst-downscaling\example\LSALST_Athens_Aug17_21.tif"
    predictors_fname = r"C:\Users\User\Dropbox\MyCodeRepository\lst-downscaling\example\eco_20180821T132843pca_pred_stack.tif"
    savedir = (
        r"C:\Users\User\Desktop\AGU19_presentation\LSALST_Athens_Aug17_21_with_resid"
    )
else:
    print("Unable to determine which OS you are using")

def main():

    Seviri_LST = gdal.Open(lst_fname)
    predictors_100m = gdal.Open(predictors_fname)
    
    data = DownscaledLST(
            LST=Seviri_LST, 
            predictors=predictors_100m,
            LST_noDataVal=-1000,
            predictors_noDataVal=-1000,
            workdir="./DLST_Example",
        )
        
    data.SetAdaBoostParams(loss="exponential", n_estimators=70)
    data.SetRandomForestParams(max_depth=9, n_estimators=50, min_samples_split=2, min_samples_leaf=1)
    data.SetElasticNetParams(l1_ratio=0.1, n_alphas=50, cv=5)
    data.SetRidgeRegrParams(alpha=1.0)
    
    # Increase the R2 threshold for discarding 
    # poorly-performing regression models.
    data.SetR2Threshold(0.4)

    data.ApplyDownscaling(residual_corr=False)
    data.SaveDLSTasGeotiff(savename="Athens_Seviri_DLST.tif")
    data.GenerateReport()

    print("LST bands that have been downscaled:")
    print(data.GetDLSTBandIndices(indexing_from_1=False))

if __name__ == "__main__":
    main()

